File: .dev/orchestrator/commands/new-adr.js
import fs from 'fs/promises'
import path from 'path'
import chalk from 'chalk'
import inquirer from 'inquirer'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

function slugify(text) {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
}

export async function newAdr(title) {
  try {
    console.log(chalk.cyan('\n--- Zuri: New ADR ---\n'))

    // Scan existing ADRs to find highest number
    const adrDir = path.join(process.cwd(), 'docs', 'decisions', 'adrs')
    await fs.mkdir(adrDir, { recursive: true })

    let maxNumber = 67 // start from 068 if none found (ADR-067 is the current highest)
    try {
      const files = await fs.readdir(adrDir)
      const adrFiles = files.filter((f) => /^ADR-\d+/.test(f) && f.endsWith('.md'))
      for (const f of adrFiles) {
        const match = f.match(/ADR-(\d+)/)
        if (match) {
          const num = parseInt(match[1], 10)
          if (num > maxNumber) maxNumber = num
        }
      }
    } catch {
      // Directory might not exist yet — maxNumber stays 67
    }

    const nextNumber = maxNumber + 1
    const paddedNumber = String(nextNumber).padStart(3, '0')

    console.log(chalk.dim(`  Next ADR number: ADR-${paddedNumber}\n`))

    // Prompt for content
    const answers = await inquirer.prompt([
      {
        type: 'editor',
        name: 'context',
        message: 'Context — what is the problem? Why do we need to decide?',
        default: '',
        waitForUseInput: false,
      },
      {
        type: 'input',
        name: 'decision',
        message: 'Decision — what was chosen and why?',
        validate: (v) => (v.trim() ? true : 'Decision is required'),
      },
      {
        type: 'input',
        name: 'consequences',
        message: 'Consequences — positive and negative impacts:',
        validate: (v) => (v.trim() ? true : 'Consequences are required'),
      },
    ])

    // Read template
    const templatePath = path.join(__dirname, '..', 'templates', 'adr.md')
    let template = await fs.readFile(templatePath, 'utf-8')

    const today = new Date().toISOString().slice(0, 10)
    const slug = slugify(title)

    // Replace placeholders
    template = template.replace(/\{\{NUMBER\}\}/g, paddedNumber)
    template = template.replace(/\{\{TITLE\}\}/g, title)
    template = template.replace(/\{\{DATE\}\}/g, today)

    // Inject context
    template = template.replace(
      '<!-- ปัญหาคืออะไร? ทำไมต้องตัดสินใจ? -->',
      `<!-- ปัญหาคืออะไร? ทำไมต้องตัดสินใจ? -->\n\n${answers.context}`
    )

    // Inject decision
    template = template.replace(
      '<!-- เลือกอะไร? ทำไม? -->',
      `<!-- เลือกอะไร? ทำไม? -->\n\n${answers.decision}`
    )

    // Inject consequences into Positive section
    template = template.replace(
      '### Positive\n- ...',
      `### Positive\n- ${answers.consequences}`
    )

    // Write ADR file
    const fileName = `ADR-${paddedNumber}-${slug}.md`
    const filePath = path.join(adrDir, fileName)
    await fs.writeFile(filePath, template, 'utf-8')

    console.log(chalk.green(`\n  Created: docs/decisions/adrs/${fileName}`))
    console.log(
      chalk.yellow(`  ADR-${paddedNumber} created — needs Boss approval\n`)
    )
  } catch (error) {
    console.error(chalk.red('[new-adr]'), error.message)
    process.exitCode = 1
  }
}

File: scripts/new-adr.py
#!/usr/bin/env python3
"""
Zuri: New ADR Creator
Usage: python scripts/new-adr.py "ADR Title"
       python scripts/new-adr.py  (interactive)
"""

import sys
import os
import re
from datetime import date
from pathlib import Path

ROOT = Path(__file__).parent.parent
ADR_DIR = ROOT / "docs" / "decisions" / "adrs"
TEMPLATE = ROOT / ".dev" / "templates" / "adr.md"


def slugify(text: str) -> str:
    return re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")


def next_adr_number() -> int:
    ADR_DIR.mkdir(parents=True, exist_ok=True)
    max_num = 68  # start from 069 if none found (ADR-068 is current highest)
    for f in ADR_DIR.iterdir():
        m = re.match(r"ADR-(\d+)", f.name)
        if m:
            max_num = max(max_num, int(m.group(1)))
    return max_num + 1


def main():
    title = " ".join(sys.argv[1:]).strip() if len(sys.argv) > 1 else ""
    if not title:
        title = input("ADR title: ").strip()
    if not title:
        print("Error: title required")
        sys.exit(1)

    num = next_adr_number()
    padded = str(num).zfill(3)
    slug = slugify(title)
    today = date.today().isoformat()

    print(f"\n  Next ADR: ADR-{padded}")
    print("  Fill in content (press Enter to leave blank):\n")

    context = input("  Context — what is the problem? Why decide? ").strip()
    decision = input("  Decision — what was chosen and why? ").strip()
    consequences = input("  Consequences — positive/negative impacts? ").strip()

    # Read template
    if TEMPLATE.exists():
        template = TEMPLATE.read_text(encoding="utf-8")
        template = template.replace("{{NUMBER}}", padded)
        template = template.replace("{{TITLE}}", title)
        template = template.replace("{{DATE}}", today)
        if context:
            template = template.replace(
                "<!-- ปัญหาคืออะไร? ทำไมต้องตัดสินใจ? -->",
                f"<!-- ปัญหาคืออะไร? ทำไมต้องตัดสินใจ? -->\n\n{context}",
            )
        if decision:
            template = template.replace(
                "<!-- เลือกอะไร? ทำไม? -->",
                f"<!-- เลือกอะไร? ทำไม? -->\n\n{decision}",
            )
        if consequences:
            template = template.replace(
                "### Positive\n- ...",
                f"### Positive\n- {consequences}",
            )
    else:
        # Fallback minimal template
        template = f"""---
title: "ADR-{padded}: {title}"
date: {today}
status: DRAFT
---

# ADR-{padded}: {title}

## Context

{context or "..."}

## Decision

{decision or "..."}

## Consequences

### Positive
- {consequences or "..."}

### Negative
- ...
"""

    filename = f"ADR-{padded}-{slug}.md"
    filepath = ADR_DIR / filename
    filepath.write_text(template, encoding="utf-8")

    print(f"\n  Created: docs/decisions/adrs/{filename}")
    print(f"  ADR-{padded} created — needs Boss approval\n")


if __name__ == "__main__":
    main()
